[inductor] Add prims._inductor_bucketize and add lowerings

davidberard98 opened the pull request:
**TL;DR**: This PR is a first step in adding lowerings for torch.bucketize. It adds an initial lowering for this op - but because this  implementation is not currently efficient, it registers the lowering for prims._inductor_bucketize. After we make the implementation more efficient, we'll remove prims._inductor_bucketize and add the lowering directly to torch.bucketize.

**Background - torch.bucketize**: torch.bucketize(values, boundaries, right=False): for an arbitrary tensor of values and a non-decreasing 1D tensor of boundaries that define buckets, it returns the index of the bucket that each of the values will fall in. e.g. for values [0, 1, 2, 3, 4] and boundaries [1, 3], it will return [0, 0, 1, 1, 2].

**Implementation**: This PR adds a new inductor op called "bucketize". In this PR it only has a triton implementation - for CPU it is a fallback. The triton implementation uses a binary search in `triton_helpers.py`. This PR also adds a new prim `_inductor_bucketize()` for testing purposes and adds lowering for this op.

~~**"right"**: The current behavior of the "right" kwarg in the inductor op is the opposite of the behavior of the torch op. "right" controls how the op treats a value that is equal to one of the boundary values. In the torch op, "right=True" means "if a value is equal to a boundary value, then put it in the bucket to the right". In the inductor op, "right=True" means "the right boundary of a bucket is closed". These are opposite. **I'm open to switching the behavior of the inductor op** - but I chose to implement this way because I think it makes more sense, and I think the torch.bucketize behavior may have been a mistake (it's the opposite of numpy.digitize).~~ Switched the behavior of the inductor bucketize op to match the torch op

* places where "right" means "if a value is equal to a boundary value, then put it in the bucket to the right" (i.e. current torch.bucketize behavior)
  + current torch.bucketize behavior
  + table in [torch.bucketize docs](https://pytorch.org/docs/stable/generated/torch.bucketize.html)
* places where "right" means "the right boundary of a bucket is closed":
  + the text description of [torch.bucketize docs](https://pytorch.org/docs/stable/generated/torch.bucketize.html) (observed in #91580)
  + [numpy.digitize](https://numpy.org/doc/stable/reference/generated/numpy.digitize.html) (which is basically the same op)

**Performance**: Benchmark script: "values" as a [16, 1024, 1024] float32 tensor and "boundaries" as a [1025] tensor (i.e. defining 1024 buckets).

As is:
```
Eager 0.30117499828338623 ms
PT2   0.9298200011253357 ms
```

But performance improves significantly if we add an additional pointwise autotuning config (WIP in #104456):
```
Eager 0.3015420138835907 ms
PT2   0.23028500378131866 ms
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78

davidberard98 committed:
[WIP][inductor] add prims.inductor_bucket_index and inductor lowerings

davidberard98 committed:
WIP - moving impl to triton_helpers.py

davidberard98 committed:
bucket_index -> bucketize

davidberard98 committed:
lint

aakhundov commented on torch/_inductor/codegen/common.py:
```diff
@@ -833,6 +841,35 @@ def reduction(name, dtype, src_dtype, reduction_type, index, value):
+                Given values (tensor) and offsets_name (reference to the name of a 1D
+                tensor), calculate the bucket that each value belongs to.
+
+                e.g. for values [-1, 0, 1, 2, 3, 4, 5, 9], offsets [0, 4, 8], right=False
+                return =        [ 0, 1, 1, 1, 1, 2, 2, 3].
```
Although this, indeed, seems to correspond to the semantics of `torch.bucketize`, in the jagged tensor context, seems that we'll need to subtract 1 from the result (as the values falling into the first bucket should have index 0).

davidberard98 replied:
In some rough benchmarks I didn't see any measurable difference in perf from subtracting the 1 vs. not having to do that (in fbgemm lowerings)

aakhundov commented on torch/_inductor/lowering.py:
```diff
@@ -1551,6 +1551,49 @@ def inner_fn(index):
     )
 
 
+@register_lowering(inductor_prims._bucketize, type_promotion_kind=None)
```
The [documentation](https://pytorch.org/docs/stable/generated/torch.bucketize.html) for `torch.bucketize` specifies that the `boundaries` "must contain a strictly increasing sequence, or the return value is undefined". In the jagged tensor context, the buckets can be zero-size (corresponding to empty sequences in the jagged tensor). I guess, our lowering here should support zero-sized buckets?

davidberard98 replied:
Good point, updated the docstring in common.py. In practice it seems like torch.bucketize obeys this behavior and we can define the behavior to work for jagged tensors.

davidberard98 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1301,6 +1323,14 @@ def reduction_resize(self, value):
         sizes[-1] = "None"
         return f"{value}[{', '.join(sizes)}]"
 
+    def reduction_size_str(self):
```
This is leftover from a bad rebase resolution I think, will remove it

davidberard98 commented on torch/_inductor/triton_helpers.py:
```diff
@@ -117,3 +117,51 @@ def _any_combine(a, b):
+    indexing_dtype,
+    right,  # bool: if true, use intervals closed on the right
+    OFFSETS_SIZE: int,
+    BLOCK_SHAPE,
```
It's actually a list/tuple. IDK if it's always 1D or if it can be multi-dimensional sometimes, so I was planning to just pass it in directly like this.

Added a comment to clarify

davidberard98 committed:
More efficient impl, cleanup, docs

davidberard98 committed:
lint

davidberard98 committed:
lint

jansel commented on torch/_inductor/lowering.py:
```diff
@@ -1556,6 +1556,49 @@ def inner_fn(index):
+):
+    assert len(boundaries.get_size()) == 1
+
+    if input.get_device().type != "cuda" or boundaries.get_device().type != "cuda":
```
Use the is_triton() helper (which does the same thing)

jansel commented on torch/_inductor/lowering.py:
```diff
@@ -1556,6 +1556,49 @@ def inner_fn(index):
+    input_loader = input.make_loader()
+
+    index_dtype = torch.int32 if out_int32 else torch.int64
+    triton_dtype = "tl.int32" if out_int32 else "tl.int64"
```
Let's not expose the triton_dytpe in the device-agnostic IR.  This will make it harder to extend to non-triton backends.

jansel commented on torch/_inductor/lowering.py:
```diff
@@ -1556,6 +1556,49 @@ def inner_fn(index):
+            boundaries.get_name(),
+            ops.index_expr(boundaries_size, index_dtype),
+            triton_dtype,
+            not right,  # ops.bucketize and torch.bucketize have opposite semantics for "right"
```
My initial feelings here are we should match the torch semantics rather than the numpy one.  That will at least make our codebase self-consistent.  I don't feel that strongly about that though.

davidberard98 committed:
Address Jason's comments

1. use is_triton()
2. pass a torch.dtype instead of a triton dtype into the bucketize inductor op
3. Switch the behavior of "right" back to torch semantics

jansel commented on torch/_inductor/triton_helpers.py:
```diff
@@ -117,3 +117,37 @@ def _any_combine(a, b):
+    values,  # 1D tensor
+    offsets_ptr,
+    indexing_dtype,
+    right,  # bool: if true, use intervals closed on the right
```
outdated comment?

davidberard98 replied:
whoops, thanks for catching that

davidberard98 committed:
fix triton_helpers comment about 'right'

davidberard98 committed:
lint

peterbell10 commented on torch/_inductor/inductor_prims.py:
```diff
@@ -71,3 +72,24 @@ def eager_force_stride(input_tensor, stride):
+
+
+_bucketize = _prims._make_prim(
+    schema="_inductor_bucketize(Tensor input, Tensor boundaries, *, bool out_int32=False, bool right=False) -> Tensor",
```
What's the advantage of adding a prim rather than just not decomposing `aten.bucketize` in inductor?

davidberard98 replied:
because this implementation is not currently efficient, I'm the lowering for prims._inductor_bucketize so that it doesn't affect perf for models that use torch.bucketize. After we make the implementation more efficient, we'll remove prims._inductor_bucketize and add the lowering directly to torch.bucketize.

peterbell10 replied:
Oh sorry for the noise, should have read the PR description more carefully.

davidberard98 commented:
@pytorchbot merge
